### ISSUE TITLE
Support for exponentiation in cvc5

### ISSUE THREAD

**Issue by yannickmoy** [cvc5/cvc5#9447]
**Describe the bug**

As currently SMT-LIB and cvc5 don't define a symbol for integer exponentiation, one has to axiomatize exponentiation, leading to limited support from the prover. For example, here is the axiomatization we use from Why3:

```smt
(declare-fun power (Int Int) Int)

;; "Power_0"
(assert (forall ((x Int)) (= (power x 0) 1)))

;; "Power_s"
(assert
  (forall ((x Int) (n Int))
    (=> (<= 0 n) (= (power x (+ n 1)) (* x (power x n))))))

;; "Power_s_alt"
(assert
  (forall ((x Int) (n Int))
    (=> (< 0 n) (= (power x n) (* x (power x (- n 1)))))))

;; "Power_1"
(assert (forall ((x Int)) (= (power x 1) x)))

;; "Power_sum"
(assert
  (forall ((x Int) (n Int) (m Int))
    (=>
      (<= 0 n)
      (=> (<= 0 m) (= (power x (+ n m)) (* (power x n) (power x m)))))))

;; "Power_mult"
(assert
  (forall ((x Int) (n Int) (m Int))
    (=> (<= 0 n) (=> (<= 0 m) (= (power x (* n m)) (power (power x n) m))))))

;; "Power_comm1"
(assert
  (forall ((x Int) (y Int))
    (=>
      (= (* x y) (* y x))
      (forall ((n Int))
        (=> (<= 0 n) (= (* (power x n) y) (* y (power x n))))))))

;; "Power_comm2"
(assert
  (forall ((x Int) (y Int))
    (=>
      (= (* x y) (* y x))
      (forall ((n Int))
        (=> (<= 0 n) (= (power (* x y) n) (* (power x n) (power y n))))))))

;; "Power_non_neg"
(assert
  (forall ((x Int) (y Int)) (=> (and (<= 0 x) (<= 0 y)) (<= 0 (power x y)))))

;; "Power_pos"
(assert
  (forall ((x Int) (y Int)) (=> (and (< 0 x) (<= 0 y)) (< 0 (power x y)))))

;; "Power_monotonic"
(assert
  (forall ((x Int) (n Int) (m Int))
    (=> (and (< 0 x) (and (<= 0 n) (<= n m))) (<= (power x n) (power x m)))))
```

This does not allow to prove a property like the following:

```smt
(declare-const a Int)
(declare-const b Int)
(declare-const e Int)

(assert (<= 0 a))
(assert (<= 0 b))
(assert (<= 0 e))

(assert (not (= (mod (power (mod a b) e) b) (mod (power a e) b))))

(check-sat)
```

That forces users to guide very heavily the proof and use multiple provers.

Are there plans to improve the support for exponentiation in cvc5?

**Command line arguments**: none 
**cvc5 version/commit**: 1.0.3
**Operating system**: Linux

**Comment by martin-cs** [comment]
If the exponent is constant then it should be pretty trivial to add support.  For non-const exponents it gets a bit tricky and IIRC it depends on if the exponents are integer, rational or real.  I think there are still some questions about decidability ( see https://en.wikipedia.org/wiki/Tarski%27s_exponential_function_problem ) and I am not sure if there are actually computationally practical algorithms for this.

**Comment by yannickmoy** [comment]
thanks @martin-cs. We're only interested in integer exponents here, like you find when you're verifying numerical algorithms on integers.

**Comment by yoni206** [comment]
Hi @yannickmoy ,

Currently we support an exponentiation operator, as long as at least one of the following two conditions is met:
1. The exponent is a constant
2. The base is 2

The operator is marked with `^`.

[Here](https://github.com/cvc5/cvc5/blob/main/test/regress/cli/regress0/arith/bug549.cvc.smt2) is an example for the first use case, and [here](https://github.com/cvc5/cvc5/blob/main/test/regress/cli/regress0/nl/pow2-pow.smt2) is an example for the second.

Making `^` more general is currently work-in-progress.

Please let us know if you have more questions.

**Comment by yannickmoy** [comment]
Thanks @yoni206, sorry I did not see your answer earlier. In the particular case I reported, none of these conditions is met, but it's interesting to know that this support exists. We'll keep it in mind to see if we can benefit from it.

**Comment by florianschanda** [comment]
@martin-cs didn't we implement integer ^ integer operations in Riposte? It was slow as shit, and obviously you need bit-vectors, but as I recall it worked.

**Comment by martin-cs** [comment]
Yes we did and yes it was.  At the risk of repeating myself, there are many different "power" functions, bit-vector, integer, real and floating-point and with constant first or second arguments.  Some of these are essentially trivial (BV, second argument constant in bit-blasting), some are decideable but horrible (floating-point, non-integral second argument), some are essentially undecideable (integer power with non-constant arguments).

**Comment by yannickmoy** [comment]
@yoni206 Returning to that issue I opened on instantiation, I tried to use int-blasting on a similar example, and that works fine, using integers are arguments of `^`, so the following is proved with int-blasting:

```smt
(set-logic AUFBVFPDTNIRA)

(declare-const x (_ BitVec 32))
(declare-const b Int)

(assert
  (not
  (=>
    (and (<= 1 b) (<= b 30))
    (= (bvudiv x ((_ int2bv 32) (^ 2 b))) (bvlshr x ((_ int2bv 32) b))))))

(check-sat)
```

Once int-blasting becomes available, we could try to rewrite our VCs to use `^` whenever possible. For example rewriting `power #x00000002 b` into `((_ int2bv 32) (mod (^ 2 (bv2nat b)) 4294967296))` instead of keeping our own `power` axiomatized uninterpreted function for the exponential function. Do you have an opinion on the benefits/drawbacks in general of doing that?

**Comment by yoni206** [comment]
Hi @yannickmoy ,

> Once int-blasting becomes available
It is now a regular options #10090 .

> the following is proved with int-blasting:
I see that in this benchmark the base is 2 (or the exponent is constant) - this is currently supported.
Are the other benchmarks similar in that respect?

> Do you have an opinion on the benefits/drawbacks in general of doing that?
It is hard to know in advance. We do have support for benchmarks like the above, but our reasoning about exponentiation is relatively basic at the moment. I guess the best bet is to try, and to share benchmarks for which this alternative is outperformed by your uninterpreted function.




`

**Comment by yannickmoy** [comment]
@yoni206 thanks for making int-blasting available as a regular option! We will try it on our testsuite.

Most cases of interest indeed have 2 or a power of 2 as base for exponentiation. We will see if we can get this rewritten to a base of 2. Experiments will definitely be easier if using int-blasting does not result in systematic errors on our generated VCs! I will let you know.

**Comment by yannickmoy** [comment]
Hi @yoni206

I tried to use exponentiation operator with a base of 2, but a property like the following cannot be proved when the upper bound on the exponent is not a small number (so 10 is already too high):

```smt
(set-logic ALL)
(set-info :smt-lib-version 2.6)

(declare-const m Int)
(declare-const n Int)

(assert (and (>= m 0) (< m 10)))
(assert (and (>= n 0) (< n 10)))

(assert (not (= (* (^ 2 m) (^ 2 n)) (^ 2 (+ m n)))))

(check-sat)
```

Is it related to the representation of the exponentiation operator? Is there a way to improve on that?

**Comment by yoni206** [comment]
Hi @yannickmoy ,

Thank you for this benchmark.
In fact, we are planning to adding this property soon. I'll let you know here once this is done.
